File: app/src/main/java/me/tagavari/airmessage/util/ContactInfo.java
package me.tagavari.airmessage.util;

import android.content.res.Resources;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;

public class ContactInfo {
	private final long identifier;
	@Nullable private final String name;
	@NonNull private final List<AddressInfo> addresses;
	
	public ContactInfo(long identifier, @Nullable String name, @NonNull List<AddressInfo> addresses) {
		this.identifier = identifier;
		this.name = name;
		this.addresses = addresses;
	}
	
	public long getIdentifier() {
		return identifier;
	}
	
	@Nullable
	public String getName() {
		return name;
	}
	
	public void addAddress(AddressInfo address) {
		addresses.add(address);
	}
	
	@NonNull
	public List<AddressInfo> getAddresses() {
		return addresses;
	}
	
	public String[] getAddressDisplayArray(Resources resources) {
		String[] displayArray = new String[addresses.size()];
		for(ListIterator<AddressInfo> iterator = addresses.listIterator(); iterator.hasNext();) {
			displayArray[iterator.nextIndex()] = iterator.next().getDisplay(resources);
		}
		return displayArray;
	}
	
	@NonNull
	@Override
	protected ContactInfo clone() {
		return new ContactInfo(identifier, name, new ArrayList<>(addresses));
	}
}

File: app/src/main/java/me/tagavari/airmessage/util/ContactInfo.kt
package me.tagavari.airmessage.util

import android.content.res.Resources
import android.net.Uri

/**
 * Represents data of a local Android contact
 * @param contactID The database identifier of this contact
 * @param name The name of this contact
 * @param thumbnailURI A URI to display the thumbnail of this contact
 * @param addresses The list of addresses associated with this contact
 */
data class ContactInfo(
	val contactID: Long,
	val name: String?,
	val thumbnailURI: Uri?,
	val addresses: List<AddressInfo>
) {
	fun getAddressDisplayList(resources: Resources): List<String> {
		return addresses.map { it.getDisplay(resources) }
	}
	
	class Builder(
		var contactID: Long
	) {
		var name: String? = null
		var thumbnailURI: Uri? = null
		var addresses: MutableList<AddressInfo> = mutableListOf()
		
		fun build() = ContactInfo(
			contactID = contactID,
			name = name,
			thumbnailURI = thumbnailURI,
			addresses = addresses
		)
	}
}